storage-proofs/porep/src/stacked/circuit/params.rs

```diff
@@ -63,57 +84,164 @@ impl<Tree: MerkleTreeTrait, G: 'static + Hasher> Proof<Tree, G> {
+        let data_leaf_num = num::AllocatedNum::alloc(cs.namespace(|| "data_leaf"), || {
+            data_leaf.ok_or_else(|| SynthesisError::AssignmentMissing)
+        })?;
+
+        // enforce inclusion of the data leaf in the tree D
+        enforce_inclusion(
             cs.namespace(|| "comm_d_inclusion"),
-            comm_d.clone(),
-            comm_d_leaf.clone(),
+            comm_d_path,
+            comm_d,
+            &data_leaf_num,
         )?;
 
-        let comm_r_last_data_leaf =
-            comm_r_last_proof.alloc_value(cs.namespace(|| "comm_r_last_data_leaf"))?;
+        // -- verify replica column openings
+
+        // Private Inputs for the DRG parent nodes.
+        let mut drg_parents = Vec::new();
+
+        for (i, parent) in drg_parents_proofs.into_iter().enumerate() {
+            let (parent_col, inclusion_path) =
+                parent.alloc(cs.namespace(|| format!("drg_parent_{}_num", i)))?;
+
+            // calculate column hash
+            let val = parent_col.hash(cs.namespace(|| format!("drg_parent_{}_constraint", i)))?;
+            // enforce inclusion of the column hash in the tree C
+            enforce_inclusion(
+                cs.namespace(|| format!("drg_parent_{}_inclusion", i)),
+                inclusion_path,
+                comm_c,
+                &val,
+            )?;
+            drg_parents.push(parent_col);
+        }
+
+        // Private Inputs for the Expander parent nodes.
+        let mut exp_parents = Vec::new();
+
+        for (i, parent) in exp_parents_proofs.into_iter().enumerate() {
+            let (parent_col, inclusion_path) =
+                parent.alloc(cs.namespace(|| format!("exp_parent_{}_num", i)))?;
+
+            // calculate column hash
+            let val = parent_col.hash(cs.namespace(|| format!("exp_parent_{}_constraint", i)))?;
+            // enforce inclusion of the column hash in the tree C
+            enforce_inclusion(
+                cs.namespace(|| format!("exp_parent_{}_inclusion", i)),
+                inclusion_path,
+                comm_c,
+                &val,
+            )?;
+            exp_parents.push(parent_col);
+        }
+
+        // -- Verify labeling and encoding
 
-        // verify encodings
-        for (layer, proof) in labeling_proofs.into_iter() {
-            let raw = replica_column_proof.c_x.get_node_at_layer(layer);
-            let labeled_node =
-                num::AllocatedNum::alloc(cs.namespace(|| format!("label_node_{}", layer)), || {
-                    raw.map(Into::into)
-                        .ok_or_else(|| SynthesisError::AssignmentMissing)
-                })?;
+        // stores the labels of the challenged column
+        let mut column_labels = Vec::new();
 
-            proof.synthesize(
-                cs.namespace(|| format!("labeling_proof_{}", layer)),
+        // PrivateInput: challenge index
+        let challenge_num = uint64::UInt64::alloc(cs.namespace(|| "challenge"), challenge)?;
```

Comment:
Why is this private? This must correspond to the public input for a given challenge, but I don't see where that is enforced. I think we either need to use the supplied public input or add a constraint enforcing equality.

Reply:
I pushed a commit which I believe addresses this issue. cc: @dignifiedquire @nicola

Reply:
yes, @porcuquine is correct - it must be a public input